lib/class-wp-block.php

```diff
@@ -0,0 +1,234 @@
+<?php
+/**
+ * Blocks API: WP_Block class
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Class representing a parsed instance of a block.
+ */
+class WP_Block implements ArrayAccess {
+
+	/**
+	 * Name of block.
+	 *
+	 * @example "core/paragraph"
+	 *
+	 * @var string
+	 */
+	public $name;
+
+	/**
+	 * Block type associated with the instance.
+	 *
+	 * @var WP_Block_Type
+	 */
+	public $block_type;
+
+	/**
+	 * Block context values.
+	 *
+	 * @var array
+	 */
+	public $context = [];
+
+	/**
+	 * All available context of the current hierarchy.
+	 *
+	 * @var array
+	 * @access protected
+	 */
+	protected $available_context;
+
+	/**
+	 * Block attribute values.
+	 *
+	 * @var array
+	 */
+	public $attributes = [];
+
+	/**
+	 * List of inner blocks (of this same class)
+	 *
+	 * @var WP_Block[]
+	 */
+	public $inner_blocks = [];
+
+	/**
+	 * Resultant HTML from inside block comment delimiters after removing inner
+	 * blocks.
+	 *
+	 * @example "...Just <!-- wp:test /--> testing..." -> "Just testing..."
+	 *
+	 * @var string
+	 */
+	public $inner_html = '';
+
+	/**
+	 * List of string fragments and null markers where inner blocks were found
+	 *
+	 * @example array(
+	 *   'inner_html'    => 'BeforeInnerAfter',
+	 *   'inner_blocks'  => array( block, block ),
+	 *   'inner_content' => array( 'Before', null, 'Inner', null, 'After' ),
+	 * )
+	 *
+	 * @var array
+	 */
+	public $inner_content = [];
+
+	/**
+	 * Constructor.
+	 *
+	 * Populates object properties from the provided block instance argument.
+	 *
+	 * The given array of context values will not necessarily be available on
+	 * the instance itself, but is treated as the full set of values provided by
+	 * the block's ancestry. This is assigned to the private `available_context`
+	 * property. Only values which are configured to consumed by the block via
+	 * its registered type will be assigned to the block's `context` property.
+	 *
+	 * @param array                  $block    Array of parsed block properties.
+	 * @param array                  $context  Optional array of ancestry context values.
```

Comment:
Yeah, the current name could definitely be cause for confusion. I do like the idea of aligning this with the property to which it's assigned internally. Something about the name "available" isn't sitting entirely well with me, which probably speaks to the property naming as well.

It's a minor thing, so I'll probably give some brief consideration, but update if I can't think of anything better.

Off the top of my head, I'm thinking something around it "coming from the ancestry", or "parent" (though "parent context" can be misinterpreted as well).

Reply:
> Maybe name this parameter `$available_context` since it ends up in `$this->available_context`?

Looking through some of the React source, I see there's a tendency there to call it "current" context. However, I think this is a case where we're not _exactly_ the same as React context, in that with React context it's a one-to-one relationship between a provider and a context.

For now, I've just updated the argument name to match the property.

See: 262b2a4